src/Interpreters/Set.cpp

```diff
@@ -331,13 +400,42 @@ ColumnPtr Set::execute(const ColumnsWithTypeAndName & columns, bool negative) co
             result = castColumnAccurate(column_to_cast, data_types[i], cast_cache.get());
         }
 
-        materialized_columns.emplace_back() = result;
-        key_columns.emplace_back() = materialized_columns.back().get();
+        // If the original column is DateTime64, check for sub-second precision
+        if (isDateTime64(column_to_cast.column->getDataType()))
+        {
+            ColumnUInt8::Ptr filtered_null_map_column = checkDateTimePrecision(column_to_cast);
+
+            // Extract existing null map and nested column from the result
+            const ColumnNullable * result_nullable_column = typeid_cast<const ColumnNullable *>(result.get());
+            const IColumn * nested_result_column = result_nullable_column
+                ? &result_nullable_column->getNestedColumn()
+                : result.get();
+
+            ColumnPtr existing_null_map_column = result_nullable_column
+                ? result_nullable_column->getNullMapColumnPtr()
+                : nullptr;
+
+            if (transform_null_in)
+            {
+                if (!null_map_holder)
+                    null_map_holder = filtered_null_map_column;
+                else
+                    null_map_holder = mergeNullMaps(null_map_holder, filtered_null_map_column);
+
+                null_map = &assert_cast<const ColumnUInt8 &>(*null_map_holder).getData();
+            }
+            else
+            {
+                ColumnPtr merged_null_map_column = mergeNullMaps(existing_null_map_column, filtered_null_map_column);
+                result = ColumnNullable::create(nested_result_column->getPtr(), merged_null_map_column);
+            }
+        }
+
+        // Append the result to materialized columns
+        materialized_columns.emplace_back(std::move(result));
+        key_columns.emplace_back(materialized_columns.back().get());
```

Comment:
Maybe we can extract it into a separate function fo readability.
But I am not sure about at.